lib/bundler/cli/exec.rb

```diff
@@ -37,10 +51,26 @@ def run
       Bundler.ui.error "bundler: command not found: #{cmd}"
       Bundler.ui.warn "Install missing gem executables with `bundle install`"
       exit 127
-    rescue ArgumentError
+    end
+
+    def kernel_load(file, *args)
+      args.pop if args.last.is_a?(Hash)
+      ARGV.replace(args)
+      ui = Bundler.ui
+      Bundler.ui = nil
+      require "bundler/setup"
+      Kernel.load(file)
+    rescue SystemExit
+      raise
+    rescue Exception => e # rubocop:disable Lint/RescueException
       Bundler.ui = ui
-      Bundler.ui.error "bundler: exec needs a command to run"
-      exit 128
+      Bundler.ui.error "bundler: failed to load command: #{cmd} (#{file})"
+      abort "#{e.class}: #{e.message}\n#{e.backtrace.join("\n  ")}"
+    end
+
+    def ruby_shebang?(file)
+      first_line = File.open(file, "rb", &:readline)
```

Comment:
probably best to only read a fixed number of bytes? an executable might not hit a `line` for a while?